File: modules/vautr-native/ios/VautrNativeModule.swift
import ExpoModulesCore
import Foundation
import uniffi_vautr_ffi

/**
 * Expo TurboModule bridging the Vautr uniffi core on iOS (VTR-061).
 *
 * Mirrors `VautrNativeBridge` (packages/vautr-client-sdk/src/mobile.ts). The
 * Rust `MobileClient` is held here; secrets never reach JS as strings —
 * `renderInOverlay`/`performAction` deliver plaintext only to the native
 * `PlatformActionHandler`.
 */
public class VautrNativeModule: Module {
    private var client: MobileClient?

    private let platformHandler = PlatformActionHandlerImpl()

    public func definition() -> ModuleDefinition {
        Name("VautrNative")

        // Native overlay handler: receives the plaintext secret for copy /
        // autofill / overlay render. JS never sees it.
        class PlatformActionHandlerImpl: PlatformActionHandler {
            func onAction(action: CoreAction, secret: String) {
                switch action {
                case .renderInOverlay:
                    // Route to the native overlay UI (released on unmount).
                    break
                case .copyToClipboard:
                    // Copy `secret` to the system pasteboard. JS never sees it.
                    UIPasteboard.general.string = secret
                case .autofill:
                    // Hand `secret` to the autofill service.
                    break
                }
            }
        }

        AsyncFunction("initialize") { (dbPath: String) in
            self.client = try await MobileClient.initialize(dbPath: dbPath)
            self.client?.setPlatformHandler(handler: self.platformHandler)
        }

        AsyncFunction("unlock") { (rawKey: Data, localGen: Double) in
            try await self.client?.unlock(rawKey: rawKey, localGen: UInt64(localGen))
        }

        AsyncFunction("listOverviews") { () -> String in
            self.client?.listOverviews() ?? "[]"
        }

        AsyncFunction("revealSecret") { (uuid: String) -> String in
            // Opaque handle (u64) surfaced to JS as a String.
            let h = try await self.client?.revealSecret(uuid: uuid) ?? 0
            return String(h)
        }

        AsyncFunction("releaseSecret") { (handle: Double) in
            self.client?.releaseSecret(handle: UInt64(handle))
        }

        AsyncFunction("renderInOverlay") { (handle: Double) in
            try await self.client?.renderSecretInOverlay(handle: UInt64(handle))
        }

        AsyncFunction("lock") {
            try await self.client?.lock()
        }

        AsyncFunction("sync") {
            try await self.client?.sync()
        }

        AsyncFunction("setSecureEnclaveBridge") {
            self.client?.setSecureEnclaveBridge(bridge: SecureEnclaveBridgeImpl())
        }
    }

    /// Real Keychain-backed SVK persistence (replaces the prior in-memory
    /// `svkBase64` placeholder). The SVK arrives as raw `Data` from the Rust
    /// core via `saveSvk` — JS never sees it as a string. Stored with
    /// `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` so it is tied to the
    /// device passcode and never leaves the device or enters an iCloud backup.
    private static let svkKeychainKey = "vautr.svk"

    private class SecureEnclaveBridgeImpl: SecureEnclaveBridge {
        func saveSvk(svk: Data) {
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrAccount as String: SecureEnclaveBridgeImpl.svkKeychainKey,
                kSecValueData as String: svk,
                kSecAttrAccessible as String: kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
            ]
            SecItemDelete(query as CFDictionary) // clear any prior value
            let status = SecItemAdd(query as CFDictionary, nil)
            if status != errSecSuccess {
                NSLog("VautrNative: saveSvk failed with status \(status)")
            }
        }

        func loadSvk() -> Data? {
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrAccount as String: SecureEnclaveBridgeImpl.svkKeychainKey,
                kSecReturnData as String: true,
                kSecMatchLimit as String: kSecMatchLimitOne,
            ]
            var item: CFTypeRef?
            let status = SecItemCopyMatching(query as CFDictionary, &item)
            guard status == errSecSuccess, let data = item as? Data else {
                return nil
            }
            return data
        }

        func deleteSvk() {
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrAccount as String: SecureEnclaveBridgeImpl.svkKeychainKey,
            ]
            SecItemDelete(query as CFDictionary)
        }

        func hasSvk() -> Bool {
            loadSvk() != nil
        }
    }
}

File: modules/vautr-native/android/src/main/java/expo/modules/vautr/VautrNativeModule.kt
package expo.modules.vautr

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import uniffi.vautr_ffi.MobileClient
import uniffi.vautr_ffi.CoreAction
import uniffi.vautr_ffi.PlatformActionHandler
import uniffi.vautr_ffi.SecureEnclaveBridge
import android.util.Base64
import android.content.Context
import android.content.SharedPreferences
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Expo TurboModule bridging the Vautr uniffi core (VTR-061).
 *
 * Each `@Func` mirrors a method on `VautrNativeBridge` (see
 * packages/vautr-client-sdk/src/mobile.ts). The Rust `MobileClient` instance is
 * held here and surfaced to JS as an opaque handle; secrets never cross into JS
 * as strings — `renderInOverlay`/`performAction` deliver plaintext only to the
 * native `PlatformActionHandler`.
 */
class VautrNativeModule : Module() {
    private var client: MobileClient? = null

    // Native overlay handler: receives the plaintext secret for copy/autofill/
    // overlay render and dispatches it to the OS clipboard / UI. JS never sees it.
    private val platformHandler = object : PlatformActionHandler {
        override fun onAction(action: CoreAction, secret: String) {
            when (action) {
                is CoreAction.RenderInOverlay -> {
                    // Show the secret in the native overlay UI. Implemented by the
                    // React Native overlay component; here we log + would route to it.
                    // The overlay is responsible for calling releaseSecret on unmount.
                }
                is CoreAction.CopyToClipboard -> {
                    // Copy `secret` to the system clipboard. JS never sees it.
                    val cm = appContext.reactContext
                        ?.getSystemService(Context.CLIPBOARD_SERVICE) as? android.content.ClipboardManager
                    cm?.setPrimaryClip(android.content.ClipData.newPlainText("vautr-secret", secret))
                }
                is CoreAction.Autofill -> {
                    // Hand `secret` to the autofill service.
                }
            }
        }
    }

    override fun definition() = ModuleDefinition {
        Name("VautrNative")

        AsyncFunction("initialize") { dbPath: String ->
            client = MobileClient.initialize(dbPath)
            client?.setPlatformHandler(platformHandler)
        }

        AsyncFunction("unlock") { rawKey: ByteArray, localGen: Double ->
            client?.unlock(rawKey, localGen.toULong())
        }

        AsyncFunction("listOverviews") { ->
            client?.listOverviews() ?: "[]"
        }

        AsyncFunction("revealSecret") { uuid: String ->
            // Returns the opaque handle (u64) as a String to JS.
            client?.revealSecret(uuid)?.toString() ?: "0"
        }

        AsyncFunction("releaseSecret") { handle: Double ->
            client?.releaseSecret(handle.toULong())
        }

        AsyncFunction("renderInOverlay") { handle: Double ->
            client?.renderSecretInOverlay(handle.toULong())
        }

        AsyncFunction("lock") {
            client?.lock()
        }

        AsyncFunction("sync") {
            client?.sync()
        }

        AsyncFunction("setSecureEnclaveBridge") {
            client?.setSecureEnclaveBridge(SecureEnclaveBridgeImpl())
        }
    }

    /// Real Android Keystore-backed SVK persistence (replaces the prior
    /// in-memory `svkBase64` placeholder). The SVK arrives as raw `ByteArray`
    /// from the Rust core via `saveSvk` — JS never sees it as a string. A
    /// non-exportable AES key in the Android Keystore encrypts the SVK; the
    /// ciphertext is stored in `EncryptedSharedPreferences` (or plain
    /// SharedPreferences on API < 23 fallback).
    private class SecureEnclaveBridgeImpl : SecureEnclaveBridge {
        private val keystoreAlias = "vautr_svk_key"
        private val prefsName = "vautr_secure"
        private val ciphertextKey = "svk"

        private fun getKeystoreKey(): SecretKey {
            val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
            ks.getKey(keystoreAlias, null)?.let { return it as SecretKey }
            val generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
            val spec = KeyGenParameterSpec.Builder(
                keystoreAlias,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setUserAuthenticationRequired(false)
                .build()
            generator.init(spec)
            return generator.generateKey()
        }

        private fun prefs(): SharedPreferences {
            val ctx = appContext.reactContext
                ?: throw IllegalStateException("VautrNative: no app context for SVK storage")
            return ctx.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
        }

        override fun saveSvk(svk: ByteArray) {
            val key = getKeystoreKey()
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, key)
            val iv = cipher.iv
            val ct = cipher.doFinal(svk)
            // Store iv || ciphertext as base64.
            val blob = Base64.encodeToString(iv + ct, Base64.DEFAULT)
            prefs().edit().putString(ciphertextKey, blob).apply()
        }

        override fun loadSvk(): ByteArray? {
            val blob = prefs().getString(ciphertextKey, null) ?: return null
            val raw = Base64.decode(blob, Base64.DEFAULT)
            val iv = raw.copyOfRange(0, 12)
            val ct = raw.copyOfRange(12, raw.size)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, getKeystoreKey(), GCMParameterSpec(128, iv))
            return cipher.doFinal(ct)
        }

        override fun deleteSvk() {
            prefs().edit().remove(ciphertextKey).apply()
        }

        override fun hasSvk(): Boolean {
            return prefs().contains(ciphertextKey)
        }
    }
}
